### PR TITLE
Add a custom token for member city

### PR DESCRIPTION
**Pull request by aaronelliotross**
Adds a custom token for member city - with fallbacks in each language.

Questions:

This code uses the member's preferred language - which @rthouvenin said wasn't so great. There's no way to get at the mailing here. 

ToDo:

- [x] Check for drift between language groups and preferred_language
- [x] Add all languages organizers / campaigners want
- [x] Get sign off that the translations and limitations are OK

### CODE REVIEW

**Comment by rthouvenin** [comment]
`preferred_language` is not ideal because a mailing may be sent in a language that does not match the preferred language. This happens when:

 - As you mention, drift between language group and language. Should not happen, but does. 
 - When someone requested to be in 2 language groups (rare, but happens)

You can access to the mailing info from the parameter `$job`: this is the id of the mailing job that the email is part of. At least that's what I understand from reading [the code of civicrm-core](https://github.com/civicrm/civicrm-core/blob/129e7e200bb451a7028c155fa7508f935d69d2fa/CRM/Mailing/BAO/Mailing.php#L1092).
You can get to the mailing through that id: 
```
    $result = civicrm_api3('MailingJob', 'get', [
      'sequential' => 1,
      'id' => $job,
      'api.Mailing.get' => ['id' => "\$value.mailing_id"],
    ]);
```
(or equivalent SQL: `SELECT * FROM civicrm_mailing_job j JOIN civicrm_mailing m ON m.id=j.mailing_id WHERE j.id=$job`)

What worries me from reading that line in core is that the hook seems to be called once per recipient. If we add 2 DB calls for each recipient, this might significantly reduce the sending speed...

File: speakcivi.php [hunk 1]
```diff
@@ -131,6 +130,85 @@ function speakcivi_civicrm_tokenValues(&$values, $cids, $job = null, $tokens = a
...
+    }
+    */
+	$dao = CRM_Utils_SQL_Select::from('civicrm_contact cc')
+		->join('ca', 'LEFT JOIN civicrm_address ca on cc.id=ca.contact_id and ca.is_primary=1')
+		->join('country', 'LEFT JOIN civicrm_country country on country.id=country_id')
+		->select('contact_id, preferred_language, city, country.iso_code')
+        ->where('cc.id IN (#cids)', array('cids' => $cids))
+        ->execute();
+
+    $FALLBACK_CITY_VALUES = array(
```
Comment (rthouvenin) [inline speakcivi.php:159]: Can you move that to CRM/Speakcivi/Tools/Dictionary.php ?
I don't like adding more stuff there, but I guess that will do until we create a proper locale file for Speakcivi.

**Comment by rthouvenin** [comment]
I wonder if that approach would work better? https://civicrm.org/blog/dave-greenberg/civimail-how-to-add-default-values-for-empty-tokens

Since there is already a token for the city, in the hook we can check whether it has a value and if not set it to the localised default.

**Comment by aaronelliotross** [comment]
Oh the default values for empty tokens thing is great! That's exactly what I was hoping to get at with the TokenProcessor version - avoiding the db call.

And for the job -> mailing (thank you for the pointers, I never would have found that!) - if job ids are sequential or otherwise unique we could cache the mapping to mailings. And cache the mailing -> language mapping. So one db call per mailing batch?

**Comment by aaronelliotross** [comment]
For what it's worth, drift seems very very minimal - at least for the languages I've checked so far. The percent of members with a preferred_language different than the language group:

.5 % - French, German, Polish
.6 % - Italian
1% - English, Spanish

From https://dataviz.wemove.eu/question/553?language=Spanish

**Comment by rthouvenin** [comment]
> And for the job -> mailing (thank you for the pointers, I never would have found that!) - if job ids are sequential or otherwise unique we could cache the mapping to mailings. And cache the mailing -> language mapping. So one db call per mailing batch?

Yes! a job belongs to a single mailing.

File: speakcivi.php [hunk 2]
```diff
@@ -131,7 +131,70 @@ function speakcivi_civicrm_tokenValues(&$values, $cids, $job = null, $tokens = a
...
+    Civi::cache()->set($cache_key, $language);
+    return $language;
+}
+
+
+function speakcivi_member_city(&$tokens, $cids, $language) {
+    $fallbacks = CRM_Speakcivi_Tools_Dictionary::fallbackCityValues();
+
+    foreach ( $cids as $contact_id ) {
+        $tokens[$contact_id]['city'] = $fallbacks[$language];
```
Comment (rthouvenin) [inline speakcivi.php:164]: Shouldn't this be set only if there is no current value? Or is it that the contact tokens are always set after the "hook tokens" and override them?

File: speakcivi.php [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 require_once 'speakcivi.civix.php';
 
+use CRM\Tools\Dictionary\CRM_Speakcivi_Tools_Dictionary as Dictionary;
```
Comment (rthouvenin) [inline speakcivi.php:5]: The change to this class was not pushed

**Review by rthouvenin — CHANGES_REQUESTED** [review]

File: speakcivi.php [hunk 4]
```diff
@@ -107,30 +109,38 @@ function speakcivi_civicrm_alterSettingsFolders(&$metaDataFolders = NULL) {
...
 function speakcivi_civicrm_tokenValues(&$values, $cids, $job = null, $tokens = array(), $context = null) {
+  $language = CRM_Speakcivi_Tools_Dictionary::findMailingLanguage($job);
+  $cityDefaultValues = CRM_Speakcivi_Tools_Dictionary::fallbackCityValues();
+  $wantsCity = array_key_exists('city', $tokens['contact']);
+
   foreach ($cids as $cid) {
     $values[$cid]['speakcivi.confirmation_hash'] = sha1(CIVICRM_SITE_KEY . $cid);
+
+    if ($wantsCity && empty($values[$cid]['city'])) {
+        $values[$cid]['city'] = $cityDefaultValues[$language];
```
Comment (rthouvenin) [inline speakcivi.php:140]: Better not assume anything about warning / error handling config, and check for existence of the key before reading `$cityDefaultValues[$language]`. You can use `CRM_Utils_Array::value($key, $array, $fallback)` for that.

File: CRM/Speakcivi/Tools/Dictionary.php [hunk 5]
```diff
@@ -453,4 +453,40 @@ public static function getMessageContent($filename, $default) {
     }
     return $content;
   }
+
+  public static function fallbackCityValues() {
+    return array(
+        "fr_FR" => "votre ville",
```
Comment (rthouvenin) [inline CRM/Speakcivi/Tools/Dictionary.php:459]: You are probably tracking this with Anne, but just so it doesn't get missed: it, es and pt (and maybe nl) are missing from the list.

Comment (aaronelliotross) [reply CRM/Speakcivi/Tools/Dictionary.php:459]: Yep - I'll update the list with whatever languages actually work.

File: speakcivi.php (hunk 4 above)
Comment (aaronelliotross) [reply speakcivi.php:140]: Updated! I didn't realize PHP had options for how to handle non-existent key - thanks!

The collections lib in Python has a subclass of dict with a default value:
 
   https://docs.python.org/3/library/collections.html#collections.defaultdict

**Review by rthouvenin — APPROVED** [review]

**Comment by rthouvenin** [comment]
Other than the list of translations, LGTM!
Whoever does the merge, please squash

**Pull request merged**